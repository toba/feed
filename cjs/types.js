"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TextType;
(function (TextType) {
    TextType["Plain"] = "text";
    TextType["HTML"] = "html";
    TextType["XHTML"] = "xhtml";
})(TextType = exports.TextType || (exports.TextType = {}));
