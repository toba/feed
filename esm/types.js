export var TextType;
(function (TextType) {
    TextType["Plain"] = "text";
    TextType["HTML"] = "html";
    TextType["XHTML"] = "xhtml";
})(TextType || (TextType = {}));
