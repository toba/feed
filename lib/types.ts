import { MimeType, LinkRelation } from '@toba/tools';

export interface ISyndicate {
   feedJSON(): Feed;
}

//export type Attributes = { [key: string]: string };
export type Attributes = Map<string, string>;

//export type AtomEntity = { [key: string]: string | Date };
export type AtomEntity = Map<string, string | Date>;

export interface Link {
   href: string;
   rel?: LinkRelation;
   type?: MimeType;
   /** Size in bytes */
   length?: number;
}

export interface Person {
   name: string;
   email?: string;
   uri?: string;
}

/**
 * Software that generated the feed.
 */
export interface Generator {
   [index: string]: string;
   name: string;
   generator?: string;
   uri?: string;
   version?: string;
}

/**
 * An Atom Feed Document is a representation of an Atom feed, including
 * metadata about the feed, and some or all of the entries associated
 * with it.  Its root is the `atom:feed` element.
 *
 * Note that there MUST NOT be any white space in a Date construct or in
 * any IRI.
 */
export interface Feed {
   id: string;
   title: string;
   subtitle: string;
   author: Person | Person[];
   link: Link;

   /**
    * If multiple `atom:entry` elements with the same `atom:id` value appear in
    * an Atom Feed Document, they represent the same entry.  Their
    * atom:updated timestamps SHOULD be different.  If an Atom Feed
    * Document contains multiple entries with the same `atom:id`, Atom
    * Processors MAY choose to display all of them or some subset of them.
    * One typical behavior would be to display only the entry with the
    * latest atom:updated timestamp.
    */
   entry: Entry[];

   rights?: string;

   /**
    * Program that generated the feed.
    */
   generator?: Generator;
}

/**
 * The `atom:entry` element represents an individual entry, acting as a
 * container for metadata and data associated with the entry.  This
 * element can appear as a child of the `atom:feed` element, or it can
 * appear as the document (i.e., top-level) element of a stand-alone
 * Atom Entry Document.
 */
export interface Entry {
   id: string;
   title: string;
   link: Link | Link[];

   /**
    * ISO-8601 value (`toISOString()` standard)
    */
   published: Date;

   /**
    * ISO-8601 value (`toISOString()` standard)
    */
   updated: Date;

   summary: string;
   rights?: string;
   category?: string | string[];
   /**
    * Author may be omitted if the `atom:feed` specifies an author.
    */
   author?: Person | Person[];
   contributor: Person | Person[];

   /** Either contents or links to the content of the entry */
   content: string;
}
