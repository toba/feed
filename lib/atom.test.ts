import '@toba/test';
import { MimeType, LinkRelation, htmlEscape } from '@toba/tools';
import {
   write,
   writeTag,
   writeTextTag,
   writeEntityTag,
   writeAttributes,
   writePerson,
   writeLink
} from './atom';
import { Link, Person, Entry, Feed, TextType } from './types';

const someDate = new Date(Date.UTC(1973, 15, 3, 12, 0, 0));
const entry: Entry = {
   id: 'id',
   title: 'title',
   summary: {
      value: '<p>summary</p>',
      type: TextType.HTML
   },
   updated: someDate,
   published: someDate,
   contributor: null,
   content: 'content',
   link: {
      href: 'http://test.com'
   }
};

test('writes basic tags', () => {
   expect(writeTag('tag', 'value1')).toBe('<tag>value1</tag>');
   expect(writeTag('when', 'value2')).toBe(`<when>value2</when>`);
});

test('writes text tags', () => {
   expect(writeTextTag('summary', entry)).toBe(
      `<summary type="html">${htmlEscape('<p>summary</p>')}</summary>`
   );
   expect(writeTextTag('title', entry)).toBe(
      '<title type="text">title</title>'
   );
});

test('writes entity tags', () => {
   const p: Person = {
      name: 'Person One'
   };
   expect(writeEntityTag('name', p)).toBe('<name>Person One</name>');
   expect(writeEntityTag('updated', entry)).toBe(
      `<updated>${someDate.toISOString()}</updated>`
   );
});

test('writes attribute key-values', () => {
   const attr = new Map<string, string>([
      ['key1', 'value1'],
      ['key2', 'value2']
   ]);
   expect(writeAttributes(attr)).toBe(' key1="value1" key2="value2"');
   expect(writeAttributes(null)).toBe('');
});

test('writes person', () => {
   const person1: Person = { name: 'Person' };
   const person2: Person = { name: 'Bob', email: 'bob@test.com' };
   const expect1 = '<author><name>Person</name></author>';
   const expect2 =
      '<author><name>Bob</name><email>bob@test.com</email></author>';

   expect(writePerson('author', person1)).toBe(expect1);
   expect(writePerson('author', person2)).toBe(expect2);
   expect(writePerson('author', [person1, person2])).toBe(expect1 + expect2);
});

test('writes link', () => {
   const href = 'http://www.test.com';
   const link1: Link = {
      href,
      rel: LinkRelation.Alternate
   };
   const link2: Link = {
      href,
      type: MimeType.Atom,
      rel: LinkRelation.Enclosure
   };
   const link3 = href;

   const expect1 = `<link href="${href}" rel="${LinkRelation.Alternate}"/>`;
   const expect2 = `<link href="${href}" rel="${
      LinkRelation.Enclosure
   }" type="${MimeType.Atom}"/>`;

   expect(writeLink(link1)).toBe(expect1);
   expect(writeLink(link2)).toBe(expect2);
   expect(writeLink([link1, link2])).toBe(expect1 + expect2);
   expect(writeLink(link3)).toBe(expect1);
});

test('writes feed', () => {
   const person: Person = { name: 'Bob', email: 'bob@test.com' };
   const feed: Feed = {
      id: 'http://feed.com',
      title: 'Feed Title',
      subtitle: 'Feed Subtitle',
      updated: someDate,
      author: person,
      contributor: person,
      link: 'http://link.com',
      entry: [entry]
   };

   expect(write(feed)).toMatchSnapshot();
});
