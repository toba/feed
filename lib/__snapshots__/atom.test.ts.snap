// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writes feed 1`] = `"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?><feed xmlns=\\"http://www.w3.org/2005/Atom\\"><id>http:&#x2F;&#x2F;feed.com</id><title>Feed Title</title><subtitle>Feed Subtitle</subtitle><author><name>Bob</name><email>bob@test.com</email></author><entry><id>id</id><title type=\\"text\\">title</title><updated>1973-03-15T00:00:00.000Z</updated><published>1973-03-15T00:00:00.000Z</published><author><name>Fred</name><email>fred@his-house.com</email></author><content type=\\"text\\">content</content><summary type=\\"html\\">&lt;p&gt;summary&lt;&#x2F;p&gt;</summary></entry><entry><id>id2</id><title type=\\"text\\">Title 2</title><updated>1973-03-15T00:00:00.000Z</updated><published>1973-03-15T00:00:00.000Z</published><content type=\\"text\\">content</content><summary type=\\"html\\">&lt;p&gt;summary&lt;&#x2F;p&gt;</summary></entry></feed>"`;
